contrib/babelfishpg_tsql/src/table_variable_mvcc.c

```diff
@@ -0,0 +1,1094 @@
+    {
+        if (HEAP_XMAX_IS_LOCKED_ONLY(tuple->t_infomask))
+            return TM_BeingModified;
+        if (HeapTupleHeaderGetCmax(tuple) >= curcid)
+            return TM_SelfModified; /* updated after scan started */
+        else
+            return TM_Invisible;    /* updated before scan started */
+    }
+
+    if (TransactionIdIsInProgress(HeapTupleHeaderGetRawXmax(tuple)))
+        return TM_BeingModified;
+
+    if (!TransactionIdDidCommit(HeapTupleHeaderGetRawXmax(tuple)))
+    {
+        /* it must have aborted unexpectedly or crashed */
+        if (find_failed_transaction(HeapTupleHeaderGetRawXmax(tuple)))
+        {
+            HeapTupleSetHintBits(tuple, buffer, HEAP_XMAX_INVALID, InvalidTransactionId);
+            return TM_Ok;
+        }
+    }
+
+    /* xmax transaction committed */
+
+    if (HEAP_XMAX_IS_LOCKED_ONLY(tuple->t_infomask))
+    {
+        HeapTupleSetHintBits(tuple, buffer, HEAP_XMAX_INVALID, InvalidTransactionId);
+        return TM_Ok;
+    }
+
+    HeapTupleSetHintBits(tuple, buffer, HEAP_XMAX_COMMITTED, HeapTupleHeaderGetRawXmax(tuple));
+    if (!ItemPointerEquals(&htup->t_self, &tuple->t_ctid))
+        return TM_Updated;      /* updated by other */
+    else
+        return TM_Deleted;      /* deleted by other */
+}
+
+
+/*
+ * HeapTupleSatisfiesVacuum
+ *
+ *  Determine the status of tuples for VACUUM purposes.  Here, what
+ *  we mainly want to know is if a tuple is potentially visible to *any*
+ *  running transaction.  If so, it can't be removed yet by VACUUM.
+ *
+ * OldestXmin is a cutoff XID (obtained from
+ * GetOldestNonRemovableTransactionId()).  Tuples deleted by XIDs >=
+ * OldestXmin are deemed "recently dead"; they might still be visible to some
+ * open transaction, so we can't remove them, even if we see that the deleting
+ * transaction has committed.
+ */
+HTSV_Result
+TVHeapTupleSatisfiesVacuum(HeapTuple htup, TransactionId OldestXmin,
+                         Buffer buffer)
+{
+    TransactionId dead_after = InvalidTransactionId;
+    HTSV_Result res;
+
+    res = TVHeapTupleSatisfiesVacuumHorizon(htup, buffer, &dead_after);
+
+    if (res == HEAPTUPLE_RECENTLY_DEAD)
+    {
+        Assert(TransactionIdIsValid(dead_after));
+
+        if (TransactionIdPrecedes(dead_after, OldestXmin))
+            res = HEAPTUPLE_DEAD;
+    }
+    else
+        Assert(!TransactionIdIsValid(dead_after));
+
+    return res;
+}
+
+
+/*
+ * Work horse for TVHeapTupleSatisfiesVacuum and similar routines.
+ *
+ * In contrast to HeapTupleSatisfiesVacuum this routine, when encountering a
+ * tuple that could still be visible to some backend, stores the xid that
+ * needs to be compared with the horizon in *dead_after, and returns
+ * HEAPTUPLE_RECENTLY_DEAD. The caller then can perform the comparison with
+ * the horizon.  This is e.g. useful when comparing with different horizons.
+ *
+ * Note: HEAPTUPLE_DEAD can still be returned here, e.g. if the inserting
+ * transaction aborted.
+ */
+HTSV_Result
```

Comment:
can we make it simpler by just detecting not alive tuples and return live for the rest. Does it matter to return various insert/delete states?
xmin is invalid
xmin with an error
xmax is valid (without an error)

Reply:
looking into this

Reply:
done